tools/shared/src/main/scala/org/scalajs/core/tools/optimizer/Analyzer.scala

```diff
@@ -350,6 +352,124 @@ final class Analyzer(semantics: Semantics, outputMode: OutputMode,
       loop(this)
     }
 
+    def tryLookupReflProxyMethod(proxyName: String): Option[MethodInfo] = {
+      if (!allowAddingSyntheticMethods) {
+        tryLookupMethod(proxyName)
+      } else {
+        /* The lookup for a target method in this code implements the
+         * algorithm defining `java.lang.Class.getMethod`. This mimics how
+         * reflective calls are implemented on the JVM, at link time.
+         *
+         * Caveat: protected methods are not ignored. This can only make an
+         * otherwise invalid reflective call suddenly able to call a protected
+         * method. It never breaks valid reflective calls. This could be fixed
+         * if the IR retained the information that a method is protected.
+         */
+
+        @tailrec
+        def loop(ancestorInfo: ClassInfo): Option[MethodInfo] = {
+          if (ancestorInfo ne null) {
+            ancestorInfo.methodInfos.get(proxyName) match {
+              case Some(m) =>
+                assert(m.isReflProxy && !m.isAbstract)
+                Some(m)
+
+              case _ =>
+                ancestorInfo.findProxyMatch(proxyName) match {
+                  case Some(target) =>
+                    val targetName = target.encodedName
+                    Some(ancestorInfo.createReflProxy(proxyName, targetName))
+
+                  case None =>
+                    loop(ancestorInfo.superClass)
+                }
+            }
```

Comment:
This lookup does not what we have done before:

> Cases like the upper currently fail on the JVM backend at runtime. The Scala.js backend uses the following rules for selection (which will also cause runtime failures):
 - If a proxy with the same signature (method name and parameters) exists in the superclass, no proxy is generated (proxy is inherited)
 - If no proxy exists in the superclass, a proxy is generated for the first method with matching signatures.

So IIUC, the following example will change call selection with this change:

```scala
class Base {
  def foo(x: Option[Int]): String = "a"
}

class Sub extends Base {
  def foo(x: Option[String]): Int = 1
}

val x: { def foo(x: Option[String]): Any } = new Sub()
x.foo(Some(1))
```

Before this change (if the doc is correct), the last line would succeed (inherited proxy). After this change, `Sub.foo` would be elected for proxy generation and the call would cause a `ClassCastException`.

Please test that whatever behavior we have afterwards is consistent with Scala JVM.